plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -449,13 +500,30 @@ extendEnv :: Name -> CekValue uni fun -> CekValEnv uni fun -> CekValEnv uni fun
 extendEnv = insertByName
 
 -- | Look up a variable name in the environment.
-lookupVarName :: forall uni fun s . (PrettyUni uni fun) => Name -> CekValEnv uni fun -> CekM s (CekValue uni fun)
+lookupVarName :: forall uni fun s . (PrettyUni uni fun) => Name -> CekValEnv uni fun -> CekM uni fun s (CekValue uni fun)
 lookupVarName varName varEnv = do
     case lookupName varName varEnv of
-        Nothing  -> throwingWithCauseExc @(CekEvaluationException uni fun) _MachineError OpenTermEvaluatedMachineError $ Just var where
+        Nothing  -> throwingWithCause _MachineError OpenTermEvaluatedMachineError $ Just var where
             var = Var () varName
         Just val -> pure val
 
+-- | Take pieces of a possibly partial builtin application and either create a 'CekValue' using
+-- 'makeKnown' or a partial builtin application depending on whether the built-in function is
+-- fully saturated or not.
+evalBuiltinApp
+    :: (GivenCekReqs uni fun s, PrettyUni uni fun)
+    => fun
+    -> Term Name uni fun ()
+    -> CekValEnv uni fun
+    -> BuiltinRuntime (CekValue uni fun)
+    -> CekM uni fun s (CekValue uni fun)
+evalBuiltinApp fun term env runtime@(BuiltinRuntime sch x cost) = case sch of
+    TypeSchemeResult _ -> do
```

Comment:
I was confused there for a bit, but now I realise that `sch` is the type scheme for the stuff that we're still waiting for, not the whole scheme for the builtin.

Reply:
... and in fact there's a comment saying something about that in the CK machine.  Maybe a similar comment here would be helpful.

Reply:
> and in fact there's a comment saying something about that in the CK machine.

Which one? I looked there, but didn't find anything.

Reply:
I've expanded the docs of `TypeScheme`:

```
-- | A 'BuiltinRuntime' represents a possibly partial builtin application.
-- We get an initial 'BuiltinRuntime' representing an empty builtin application (i.e. just the
-- builtin with no arguments) by instantiating (via 'toBuiltinRuntime') a 'BuiltinMeaning'.
--
-- A 'BuiltinRuntime' contains info that is used during evaluation:
--
-- 1. the 'TypeScheme' of the uninstantiated part of the builtin. I.e. initially it's the type
--      scheme of the whole builtin, but applying or type-instantiating the builtin peels off
--      the corresponding constructor from the type scheme
-- 2. the (possibly partially instantiated) denotation
-- 3. the (possibly partially instantiated) costing function
--
-- All the three are in sync in terms of partial instantiatedness due to 'TypeScheme' being a
-- GADT and 'FoldArgs' and 'FoldArgsEx' operating on the index of that GADT.
data BuiltinRuntime term =
    forall args res. BuiltinRuntime
        (TypeScheme term args res)
        (FoldArgs args res)  -- Must be lazy, because we don't want to compute the denotation when
                             -- it's fully saturated before figuring out what it's going to cost.
        (FoldArgsEx args)    -- We make this lazy, so that evaluators that don't care about costing
                             -- can put @undefined@ here. TODO: we should test if making this strict
                             -- introduces any measurable speedup.
```

and added a reference to them from `CekValue`: